pymeasure/instruments/hp/hp3478A.py

```diff
@@ -0,0 +1,490 @@
+    MODES={"DCV": "F1",
+           "ACV": "F2",
+           "R2W": "F3",
+           "R4W": "F4",
+           "DCI": "F5",
+           "ACI": "F6",
+           "Rext": "F7",
+           }
+
+    RANGES={"DCV": {3E-2: "R-2", 3E-1: "R-1",3: "R0",30: "R1",300:"R2",
+                    "auto": "RA"},
+            "ACV": {3E-1: "R-1", 3: "R0",30: "R1" ,300: "R2","auto": "RA"},
+            "R2W": {30: "R1", 300: "R2", 3E3: "R3",3E4: "R4",3E5: "R5",
+                    3E6: "R6", 3E7: "R7","auto": "RA"},
+            "R4W": {30: "R1",300: "R2", 3E3: "R3",3E4: "R4",3E5: "R5",
+                    3E6: "R6", 3E7: "R7","auto": "RA"},
+            "DCI": {3E-1: "R-1",3: "R0","auto": "RA"},
+            "ACI": {3E-1: "R-1",3: "R0","auto": "RA"},
+            "Rext": {3E7: "R7","auto": "RA"},
+            }
+
+    TRIGGERS={
+            "auto": "T1",
+            "internal": "T1",
+            "external": "T2",
+            "single": "T3",
+            "hold": "T4",
+            "fast": "T5",
+            }
+
+    ERRORS={
+            1: "Calibration error",
+            2: "RAM error",
+            4: "ROM error",
+            8: "AD slope error",
+            16: "AD converter error",
+            32: "AD link error",
+           }
+
+
+
+    #decoder functions
+    @staticmethod
+    def decode_status(status_bytes, field=None):
+        """Method to handle the decoding of the status bytes into something meaningfull
+
+        :param status_bytes:    list of bytes to be decoded
+        :return ret_val:
+
+        """
+
+        ret_val = Status(Status_bytes(*status_bytes.encode(encoding="ASCII")))
+        if field is None:
+            return ret_val.b
+        elif field == "SRQ":
+            return getattr(ret_val.B, "byte3")
+        else:
+            return getattr(ret_val.b, field)
+
+    @classmethod
+    def decode_mode(cls,status_bytes):
+        """Method to decode current mode
+
+        :param status_bytes:   list of bytes to be decoded
+        :return cur_mode: string with the current trigger mode
+        :rtype cur_mode: str
+
+        """
+        cur_stat = Status(Status_bytes(*status_bytes.encode(encoding="ASCII")))
+        function = cur_stat.b.function
+        inv_modes = {v: k for k, v in cls.MODES.items()}
+        cur_mode = inv_modes["F"+str(function)]
+        return cur_mode
+
+    @classmethod
+    def decode_range(cls,status_bytes):
+        """Method to decode current range
+
+        :param status_bytes:   list of bytes to be decoded
+        :return cur_range: string with the current trigger mode
+        :rtype cur_range: float
+
+        """
+        cur_stat = Status(Status_bytes(*status_bytes.encode(encoding="ASCII")))
+        function = cur_stat.b.function
+        inv_modes = {v: k for k, v in cls.MODES.items()}
+        cur_mode = inv_modes["F"+str(function)]
+        rnge = cur_stat.b.range
+        cur_range = list(cls.RANGES[cur_mode].keys())[rnge-1]
+        return cur_range
```

Comment:
I think the statement `list(cls.RANGES[cur_mode].keys())[rnge-1]` rely on the fact that dictionary keys are ordered which is not the case. You could try to use `OrderedDict `or some other method, I believe.

Reply:
@msmttchr Thank for the reminder about the fact that the order of elements in dicts can not always be guranteed.
Probably because I am programming here on my ArchLinux machine with Python 3.9.x, I never encountered a problem with the dicts.

Over the next days Ill look at this range read out method again, I think ther emus be a better method to decuct the current range form the status bits, without the reversed dict.

Reply:
I'm not sure what the best approach is. However, we've had this discussion recently for another PR as well; there in the end it was decided it would be safe to rely on the order of a dict. See https://github.com/pymeasure/pymeasure/pull/276#issuecomment-742036301_

Reply:
indeed, assuming order should be safe enough until 3.6 drops off the list of supported versions at EOL.

Reply:
Thanks for the comments reading the order of the dict elements.
While testing the range method in more detail I found a nasty issue:
If the instrument is in DC or AC current measruement mode the first byte of the status bytes is >128, and therefore I get an error from the instrument.measurement method:
```
meter.measure_ACI
Out[2]: 0.00122

meter.range
Traceback (most recent call last):

  File "<ipython-input-3-09f873ffe539>", line 1, in <module>
    meter.range

  File "/home/robby/builds/pymeasure/pymeasure/instruments/instrument.py", line 235, in fget
    vals = self.values(command_process(get_command), **kwargs)

  File "/home/robby/builds/pymeasure/pymeasure/instruments/instrument.py", line 130, in values
    return self.adapter.values(command, **kwargs)

  File "/home/robby/builds/pymeasure/pymeasure/adapters/adapter.py", line 87, in values
    results = str(self.ask(command)).strip()

  File "/home/robby/builds/pymeasure/pymeasure/adapters/visa.py", line 112, in ask
    return self.connection.query(command)

  File "/home/robby/.local/lib/python3.9/site-packages/pyvisa/resources/messagebased.py", line 644, in query
    return self.read()

  File "/home/robby/.local/lib/python3.9/site-packages/pyvisa/resources/messagebased.py", line 486, in read
    message = self._read_raw().decode(enco)

UnicodeDecodeError: 'ascii' codec can't decode byte 0xc6 in position 0: ordinal not in range(128)
```
I still need to do a more deep  investigation, but I might rework the range method to use a more low-level read method (e.g. pyvisas read_raw).

Reply:
I think I found the reason for this:
If I switch the instrument to AC-Current measurment (with 4 digit resolution) the first status byte has the value  0xc6(=198).
(In DC-Current measurment the value is 0xa6.)
As this value is 198 in decimal, it is bigger then the 128 allowed by the adatper.read() called adapter.ask().

for now I am working on a staticmethod get_status which communciates to the instrument and reads the 5 bytes of status bytes.

Reply:
Adapters have a binary_values method that might be useful, although it returns a numpy array and not the raw bytes. maybe we should think about exposing a generic read_binary_values in correspondence to the write_binary_values that we already have.

Reply:
@bilderbuchi I like your idea,
I have now reworked my code with a get_status function using the read_bytes() method.